fluent-syntax/src/parser/mod.rs

```diff
@@ -1,14 +1,842 @@
+                {
+                    if text_element_type == TextElementType::NonBlank {
+                        last_non_blank = Some(elements.len());
+                    }
+                    elements.push(PatternElementPointers::TextElement(
+                        slice_start,
+                        end,
+                        indent,
+                        text_element_role,
+                    ));
+                }
+            }
+
+            text_element_role = match termination_reason {
+                TextElementTermination::LineFeed => TextElementPosition::LineStart,
+                TextElementTermination::CarriageReturn => TextElementPosition::Continuation,
+                TextElementTermination::PlaceableStart => TextElementPosition::Continuation,
+                TextElementTermination::EOF => TextElementPosition::Continuation,
+            };
+        }
+    }
+
+    if let Some(last_non_blank) = last_non_blank {
+        let elements = elements
+            .into_iter()
+            .take(last_non_blank + 1)
+            .enumerate()
+            .filter_map(|(i, elem)| match elem {
+                PatternElementPointers::Placeable(exp) => Some(ast::PatternElement::Placeable(exp)),
+                PatternElementPointers::TextElement(start, end, indent, role) => {
+                    let start = if role == TextElementPosition::LineStart {
+                        if let Some(common_indent) = common_indent {
+                            start + cmp::min(indent, common_indent)
+                        } else {
+                            start + indent
+                        }
+                    } else {
+                        start
+                    };
+                    let slice = ps.get_slice(start, end);
+                    if last_non_blank == i {
+                        if slice == "\n" {
+                            return None;
+                        }
+                        Some(ast::PatternElement::TextElement(slice.trim_end()))
+                    } else {
+                        Some(ast::PatternElement::TextElement(slice))
+                    }
+                }
+            })
+            .collect();
+        return Ok(Some(ast::Pattern { elements }));
+    }
+
+    Ok(None)
+}
+
+fn get_text_slice<'p>(
+    ps: &mut ParserStream<'p>,
+) -> Result<(usize, usize, TextElementType, TextElementTermination)> {
+    let start_pos = ps.ptr;
+    let mut text_element_type = TextElementType::Blank;
+
+    while ps.ptr < ps.length {
+        match ps.source[ps.ptr] {
+            b' ' => ps.ptr += 1,
+            b'\n' => {
+                ps.ptr += 1;
+                return Ok((
+                    start_pos,
+                    ps.ptr,
+                    text_element_type,
+                    TextElementTermination::LineFeed,
+                ));
+            }
+            b'\r' if ps.is_byte_at(b'\n', ps.ptr + 1) => {
```

Comment:
Can this go out-of-bounds for a stream ending in `\r`? No idea what happens in rust if that happens.

Reply:
I think this is OK. `ParserStream::is_byte_at` uses the [checked `String::get` method](https://doc.rust-lang.org/std/string/struct.String.html#method.get) which returns `None` whenever the unchecked one would panic.